File: Leet Code Solutions/863. All Nodes Distance K in Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> distanceK(TreeNode* root, TreeNode* target, int K) {
        unordered_map<TreeNode*, vector<TreeNode*>> graph;
        buildGraph(root, nullptr, graph);
        queue<pair<TreeNode*, int>> q;
        unordered_set<TreeNode*> visited;
        vector<int> result;
        
        q.push({target, 0});
        visited.insert(target);
        
        while (!q.empty()) {
            TreeNode* node = q.front().first;
            int distance = q.front().second;
            q.pop();
            
            if (distance == K) {
                result.push_back(node->val);
            }
            
            if (distance > K) {
                break;
            }
            
            for (TreeNode* neighbor : graph[node]) {
                if (visited.find(neighbor) == visited.end()) {
                    visited.insert(neighbor);
                    q.push({neighbor, distance + 1});
                }
            }
        }
        
        return result;
    }
    
    void buildGraph(TreeNode* node, TreeNode* parent, unordered_map<TreeNode*, vector<TreeNode*>>& graph) {
        if (node == nullptr) {
            return;
        }
        
        if (graph.find(node) == graph.end()) {
            graph[node] = vector<TreeNode*>();
        }
        
        if (parent != nullptr) {
            graph[node].push_back(parent);
            graph[parent].push_back(node);
        }
        
        buildGraph(node->left, node, graph);
        buildGraph(node->right, node, graph);
    }
};

File: Leet Code Solutions/863. All Nodes Distance K in Binary Tree.java
class Solution {
    public List<Integer> distanceK(TreeNode root, TreeNode target, int k) {
        List<Integer> ans = new ArrayList<>();
        Map<Integer, TreeNode> parent = new HashMap<>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode top = queue.poll();

                if (top.left != null) {
                    parent.put(top.left.val, top);
                    queue.offer(top.left);
                }

                if (top.right != null) {
                    parent.put(top.right.val, top);
                    queue.offer(top.right);
                }
            }
        }

        Map<Integer, Integer> visited = new HashMap<>();
        queue.offer(target);
        while (k > 0 && !queue.isEmpty()) {
            int size = queue.size();

            for (int i = 0; i < size; i++) {
                TreeNode top = queue.poll();

                visited.put(top.val, 1);

                if (top.left != null && !visited.containsKey(top.left.val)) {
                    queue.offer(top.left);
                }

                if (top.right != null && !visited.containsKey(top.right.val)) {
                    queue.offer(top.right);
                }

                if (parent.containsKey(top.val) && !visited.containsKey(parent.get(top.val).val)) {
                    queue.offer(parent.get(top.val));
                }
            }

            k--;
        }

        while (!queue.isEmpty()) {
            ans.add(queue.poll().val);
        }
        return ans;
    }
}
